docs/agent-vs-aea.md

```diff
@@ -0,0 +1,238 @@
+
+First, import the python and application specific libraries.
+``` python
+import os
+import time
+from threading import Thread
+from typing import List, Optional
+
+from aea.agent import Agent
+from aea.configurations.base import PublicId
+from aea.connections.base import Connection
+from aea.connections.stub.connection import StubConnection
+from aea.crypto.default import DEFAULT
+from aea.crypto.helpers import DEFAULT_PRIVATE_KEY_FILE, _create_default_private_key
+from aea.crypto.wallet import Wallet
+from aea.mail.base import Envelope
+```
+
+Unlike an `AEA`, an `Agent` does not require a `LedgerApis` or `Resources` module.
+
+However, we need to implement 5 abstract methods:
+- `setup()`
+- `act()`
+- `react()`
+- `update()`
+- `teardown()`
+
+
+When we run an agent, `start()` calls `setup()` and then the main agent loop. The main agent loop calls `act()`, `react()` and `update()` on each tick. When the agent is stopped via `stop()` then `teardown()` is called.
+
+Such a lightweight agent can be used to implement simple logic.
+
+## Code an `Agent`
+
+We define our `Agent` which simply receives envelopes, prints the sender address and protocol_id and returns it unopened.
+```python
+INPUT_FILE = "input.txt"
+OUTPUT_FILE = "output.txt"
+
+
+class MyAgent(Agent):
+
+    def __init__(self, name: str, connections: List[Connection], wallet: Wallet):
+        super().__init__(name, connections, wallet)
+        self.tick = 0
+
+    def setup(self):
+        pass
+
+    def act(self):
+        print("Act called for tick {}".format(self.tick))
+
+    def react(self):
+        print("React called for tick {}".format(self.tick))
+        while not self.inbox.empty():
+            envelope = self.inbox.get_nowait()  # type: Optional[Envelope]
+            if envelope is not None:
+                sender = envelope.sender
+                receiver = envelope.to
+                envelope.to = sender
+                envelope.sender = receiver
+                print("Received envelope from {} with protocol_id={}".format(sender, envelope.protocol_id))
+                self.outbox.put(envelope)
+
+    def update(self):
+        print("Update called for tick {}".format(self.tick))
+        self.tick += 1
+
+    def teardown(self):
+        pass
+```
+
+## Instantiate an `Agent`
+
+``` python
+# Ensure the input and output files do not exist initially
+if os.path.isfile(INPUT_FILE):
+    os.remove(INPUT_FILE)
+if os.path.isfile(OUTPUT_FILE):
+    os.remove(OUTPUT_FILE)
+
+# Create a private key
+_create_default_private_key()
+
+# Set up the wallet, stub connection, ledger and (empty) resources
+wallet = Wallet({DEFAULT: DEFAULT_PRIVATE_KEY_FILE})
+stub_connection = StubConnection(input_file_path=INPUT_FILE, output_file_path=OUTPUT_FILE, connection_id=PublicId("fetchai", "stub", "0.1.0"))
```

Comment:
@marcofavorito I don't understand how this is an improvement over the previous API that we now need to pass the Public id into the connection. In my opinion, the StubConnection defines the public id, so i should not need to assign it...

Reply:
This issue only emerges in the programmatic initialization of the connections.

The problem it tries to solve is the following. Let's say you want to customize the `StubConnection` and want its public id to be `david/stub:0.1.0`. You can literally copy-and-paste the code and just change the field 'author' of the `connection.yaml`. Earlier, instead, you had to change the part of the code that hard-codes the public id.

Both approaches can both live though. Nothing prevents us in `StubConnection.__init__` to have the call to `super().__init__` to include the hard-coded public id:
- with the programmatic initialization, the user does not care about the connection id (as long as the default value makes sense for his purposes)
- nothing changes for the initialization on the framework side (that is when we use `aea run`).

Reply:
## TL;DR

add `super().__ini__(connection_id=PublicId("fetchai", "stub", "0.1.0")` in the constructor of `StubConnection`.

Reply:
Ok, not quite sure I agree with the idea that you'd want to allow for the above change of id with same underlying code but your quick fix works for me so I'll add that for now. To me, the `public_id` is very closely related to the code. In the future i'd like to have a one-to-one mapping between the `public_id` and an IPFS-like hash of the underlying implementation. That way, we can get higher integrity across the stack. I.e. an agent can use the public_id, recover the hash from a securely distributed lookup table, then fetch the package from the registry, initialize it and use it. So it makes sense then that the public id is hard coded in the object. What do you think?

Reply:
yes, we should definitely move toward that direction eventually (one-to-one mapping).

However, I think in the short term the current approach is less error-prone. Otherwise, we would have had to guarantee consistency between the configuration file `connection.yaml` and hard-coded public id in the Connection class. Instead, now the information is only in one place: `connection.yaml`.

Reply:
> you'd want to allow for the above change of id with same underlying code 

of course, it is not recommended to edit anything inside `vendor/`. But what I'd do if I were a user and wanted to slightly modify the implementation of say `StubConnection` is to copy it in custom `connections/`.

Reply:
Yeah sure, but then you could change the hard coded id as well :). I think what you say wrt to connection.yaml and connection.py consistency is key. For this hashing will help a lot. We can include an IPFS hash of all files but the `connection.yaml` file in a given folder the `connection.yaml`. Then at load time we check that the files are consistent. Then we have a second hash of the entire directory which will be used for the `public_id` to `hash` mapping. Thoughts?